From a2364d6afe3f3eb8b10a50cdda97c3770acb1100 Mon Sep 17 00:00:00 2001
From: flawedworld <flawedworld@flawed.world>
Date: Fri, 15 Oct 2021 17:07:13 +0100
Subject: [PATCH] enable secondary user logout support by default

Ported from 11, 12 moved the isLogoutEnabled boolean to ActiveAdmin.java

squash 663f066ddece fix DevicePolicyManager#logoutUser() never succeeding

fix DevicePolicyManager#logoutUser() never succeeding

To succeed, userId to switch to needs to be set with setLogoutUserIdLocked(), but this is not done
in both callers of this method (both of which are "End session" buttons), making them no-ops.
---
 .../java/com/android/server/devicepolicy/ActiveAdmin.java | 2 +-
 .../server/devicepolicy/DevicePolicyManagerService.java   | 8 +++++++-
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/services/devicepolicy/java/com/android/server/devicepolicy/ActiveAdmin.java b/services/devicepolicy/java/com/android/server/devicepolicy/ActiveAdmin.java
index dc8cec91001b..2ae616fc9cc5 100644
--- a/services/devicepolicy/java/com/android/server/devicepolicy/ActiveAdmin.java
+++ b/services/devicepolicy/java/com/android/server/devicepolicy/ActiveAdmin.java
@@ -233,7 +233,7 @@ class ActiveAdmin {
     boolean requireAutoTime = false;
     boolean forceEphemeralUsers = false;
     boolean isNetworkLoggingEnabled = false;
-    boolean isLogoutEnabled = false;
+    boolean isLogoutEnabled = true;
 
     // one notification after enabling + one more after reboots
     static final int DEF_MAXIMUM_NETWORK_LOGGING_NOTIFICATIONS_SHOWN = 2;
diff --git a/services/devicepolicy/java/com/android/server/devicepolicy/DevicePolicyManagerService.java b/services/devicepolicy/java/com/android/server/devicepolicy/DevicePolicyManagerService.java
index 3d9522ee9733..a3fbd93bf5ff 100644
--- a/services/devicepolicy/java/com/android/server/devicepolicy/DevicePolicyManagerService.java
+++ b/services/devicepolicy/java/com/android/server/devicepolicy/DevicePolicyManagerService.java
@@ -12759,6 +12759,12 @@ public class DevicePolicyManagerService extends IDevicePolicyManager.Stub {
         Preconditions.checkCallAuthorization(canManageUsers(caller)
                 || hasCallingOrSelfPermission(permission.INTERACT_ACROSS_USERS));
 
+        synchronized (getLockObject()) {
+            if (getLogoutUserIdUnchecked() == UserHandle.USER_NULL) {
+                setLogoutUserIdLocked(UserHandle.USER_SYSTEM);
+            }
+        }
+
         int currentUserId = getCurrentForegroundUserId();
         if (VERBOSE_LOG) {
             Slogf.v(LOG_TAG, "logout() called by uid %d; current user is %d", caller.getUid(),
@@ -19223,7 +19229,7 @@ public class DevicePolicyManagerService extends IDevicePolicyManager.Stub {
     @Override
     public boolean isLogoutEnabled() {
         if (!mHasFeature) {
-            return false;
+            return true;
         }
         synchronized (getLockObject()) {
             ActiveAdmin deviceOwner = getDeviceOwnerAdminLocked();
-- 
2.45.2

